README.md
# Desafio

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-05-12
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-05-12
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/desafio1/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/desafio1/compare/0.1.0...0.1.1


test/desafio1/core_test.clj
(ns desafio1.core-test
  (:require [clojure.test :refer :all]
            [desafio1.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/desafio/db.clj
(ns desafio.db
  (:require [datomic.api :as datomic]))

(def db-uri "datomic:dev://localhost:4334/desafio")

(defn abre-conexao! []
  (datomic/create-database db-uri)
  (datomic/connect db-uri))

(defn apaga-banco! []
  (datomic/delete-database db-uri))

(def schema [; Cartão de crédito
             {:db/ident :card/id
              :db/valueType :db.type/uuid
              :db/cardinality :db.cardinality/one
              :db/unique :db.unique/identity
              :db/doc "uuid do cartão"}
             {:db/ident :card/number
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "número do cartão"}
             {:db/ident :card/cvv
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "código de segurança do cartão"}
             {:db/ident :card/due-date
              :db/valueType :db.type/instant
              :db/cardinality :db.cardinality/one
              :db/doc "Data de validade do cartão"}
             {:db/ident :card/limit
              :db/valueType :db.type/bigdec
              :db/cardinality :db.cardinality/one
              :db/doc "Limite do cartão"}

             ; Usuário
             {:db/ident :client/id
              :db/valueType :db.type/uuid
              :db/cardinality :db.cardinality/one
              :db/unique :db.unique/identity
              :db/doc "uuid do cliente"}
             {:db/ident :client/name
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "nome do cliente"}
             {:db/ident :client/email
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "email do cliente"}
             {:db/ident :client/cpf
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "CPF do cliente"}
             {:db/ident :client/card
              :db/valueType :db.type/ref
              :db/cardinality :db.cardinality/many
              :db/doc "cartão do cliente"}

             ; Purchase
             {:db/ident :purchase/id
              :db/valueType :db.type/uuid
              :db/cardinality :db.cardinality/one
              :db/unique :db.unique/identity
              :db/doc "uuid da compra"}
             {:db/ident :purchase/occuredAt
              :db/valueType :db.type/instant
              :db/cardinality :db.cardinality/one
              :db/doc "data da compra"}
             {:db/ident :purchase/value
              :db/valueType :db.type/bigdec
              :db/cardinality :db.cardinality/one
              :db/doc "valor da compra"}
             {:db/ident :purchase/place
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "local da compra"}
             {:db/ident :purchase/card
              :db/valueType :db.type/ref
              :db/cardinality :db.cardinality/one
              :db/doc "cartão utilizado na compra"}
             {:db/ident :purchase/category
              :db/valueType :db.type/ref
              :db/cardinality :db.cardinality/one
              :db/doc "categoria da compra"}

             ; Category
             {:db/ident :category/id
              :db/valueType :db.type/uuid
              :db/cardinality :db.cardinality/one
              :db/unique :db.unique/identity
              :db/doc "uuid da categoria"}
             {:db/ident :category/name
              :db/valueType :db.type/string
              :db/cardinality :db.cardinality/one
              :db/doc "nome da categoria"}])

(defn cria-schema! [conn]
  (datomic/transact conn schema))

(defn link-card-to-user! [conn client card]
  (datomic/transact conn [[:db/add
                          [:client/id (:client/id client)]
                          :client/card [:card/id (:card/id card)]]]))

(defn link-category-to-purchase! [conn purchase category]
  (datomic/transact conn [[:db/add
                          [:purchase/id (:purchase/id purchase)]
                          :purchase/category [:category/id (:category/id category)]]]))

(defn link-card-to-purchase! [conn purchase card]
  (datomic/transact conn [[:db/add
                          [:purchase/id (:purchase/id purchase)]
                          :purchase/card [:card/id (:card/id card)]]]))

(defn create-client! [conn client card] 
  (datomic/transact conn [client card])
  (link-card-to-user! conn client card))

(defn create-purchase! [conn purchase card category]
  (datomic/transact conn [purchase])
  (link-category-to-purchase! conn purchase category)
  (link-card-to-purchase! conn purchase card))

(defn create-categories! [conn categories]
  (datomic/transact conn categories))

(defn todas-as-compras [conn]
  (datomic/q '[:find (pull ?purchase [*])
               :where [?purchase :purchase/id]]
             (datomic/db conn)))

(defn count-clients-purchases [conn]
  (datomic/q '[:find (count ?purchase) (pull ?client [*])
               :where
               [?card :card/id]
               [?purchase :purchase/card ?card]
               [?client :client/card ?card]
               :keys count client]
             (datomic/db conn)))

(defn top-purchases-count-client [conn]
  (->>
   (datomic/q '[:find (count ?purchase) (pull ?client [*])
               :where
               [?card :card/id]
               [?purchase :purchase/card ?card]
               [?client :client/card ?card]
               :keys count client]
             (datomic/db conn))
   (apply max-key :count)))

(defn top-purchases-value-client [conn]
  (datomic/q '[:find ?value
               (pull ?card [:card/id {:client/_card [*]}])
               :where
               [(q '[:find (max ?value)
                     :where
                     [?purchase :purchase/value ?value]]
                   $) [[?value]]]
               [?purchase :purchase/value ?value]
               [?purchase :purchase/card ?card]
               :keys value client]
             (datomic/db conn)))

(defn client-without-purchase [conn]
  (datomic/q '[:find (pull ?client [*])
               :where
               [?card :card/id]
               (not-join [?card]
                [?purchase :purchase/card ?card])
               [?client :client/card ?card]
               ]
             (datomic/db conn)))


src/desafio/desafio1.clj
(ns desafio.desafio1)
(require '[clj-time.core :as t])

(defn cards
  []
  [{:number "1111111111111111"
    :cvv 123
    :due-date (t/date-time 2022 12 30)
    :limit 1000}
   {:number "2222222222222222"
    :cvv 321
    :due-date (t/date-time 2024 06 18)
    :limit 1500}
   {:number "3333333333333333"
    :cvv 231
    :due-date (t/date-time 2025 04 30)
    :limit 2500}])

(defn clients
  []
  [{:name "Felipe Melo"
    :cpf "11111111111"
    :email "felipe.bezerra@nubank.com.br"
    :card (get cards 0)}
   {:name "Bianca Abreu"
    :cpf "22222222222"
    :email "bianca.abreu@nubank.com.br"
    :card (get cards 1)}
   {:name "Lionel Messi"
    :cpf "33333333333"
    :email "lionel.messi@nubank.com.br"
    :card (get cards 2)}])

(defn purchases
  []
  [{:date (t/date-time 2021 05 10)
    :value 50.64
    :place "mercado"
    :category "food"
    :card (get cards 0)}
   {:date (t/date-time 2021 05 11)
    :value 234.9
    :place "steam"
    :category "entreteniment"
    :card (get cards 0)}
   {:date (t/date-time 2021 05 06)
    :value 4.9
    :place "net shoes"
    :category "sport"
    :card (get cards 1)}
   {:date (t/date-time 2021 04 25)
    :value 42.94
    :place "net shoes"
    :category "sport"
    :card (get cards 2)}])

(defn filter-purchase
  [key, value]
  (filter #(= value (key %)) (purchases)))

(defn filter-client
  [key, value]
  (filter #(= value (key %)) (clients)))

(defn values-by-category
  [category]
  (let [filtered-values (filter-purchase :category category)]
    (reduce + (map :value filtered-values))))

(defn month-invoice
  [cpf]
  (let [client (filter-client :cpf cpf)
        client-purchases (filter-purchase #(-> % :card :number) (:number (:card client)))]
    (->>
     client-purchases
     (filter #(= (t/month (:date %)) (t/month (t/now))))
     (map :value)
     (reduce +))))

(println (values-by-category "sport"))
(println (filter-purchase :value 4.9))
(println (filter-purchase :place "net shoes"))

(month-invoice "11111111111")


src/desafio/desafio2.clj
(ns desafio.desafio2
  (:use clojure.pprint)
  (:require [desafio.db :as db])
  (:require [desafio.models :as models]))
(use 'clojure.instant)

(def conn (db/abre-conexao!))
(db/cria-schema! conn)

; ---------------- cria clientes com cartões ----------------
(def client1 (models/new-client "Felipe Melo" "11111111111" "felipe@nubank.com.br"))
(def card1 (models/new-card "1111111111111111" "123" (read-instant-date "2022-12-30T10:22:22") 1000M))

(db/create-client! conn client1 card1)

(def client2 (models/new-client "Bianca Abreu" "22222222222" "bianca@nubank.com.br"))
(def card2 (models/new-card "2222222222222222" "321" (read-instant-date "2024-06-18T10:22:22") 2000M))

(db/create-client! conn client2 card2)

(def client3 (models/new-client "Marisa Silva" "33333333333" "marisa@nubank.com.br"))
(def card3 (models/new-card "3333333333333333" "231" (read-instant-date "2025-04-30T10:22:22") 3500M))

(db/create-client! conn client3 card3)

(def client4 (models/new-client "James" "44444444444" "james@nubank.com.br"))
(def card4 (models/new-card "4444444444444444" "234" (read-instant-date "2025-04-30T10:22:22") 35000M))

(db/create-client! conn client4 card4)
; ---------------- cria clientes com cartões ----------------

; --------------------- cria categogiras --------------------
(def category1 (models/new-category "mercado"))
(def category2 (models/new-category "serviço"))
(def category3 (models/new-category "eletrônico"))

(db/create-categories! conn [category1 category2 category3])
; --------------------- cria categogiras --------------------

; -------------------- criação de compras -------------------
(def purchase1 (models/new-purchase (read-instant-date "2021-05-20T10:22:22") 240M "Rio de Janeiro"))
(def purchase2 (models/new-purchase (read-instant-date "2021-05-20T10:22:22") 7240M "Rio de Janeiro"))
(def purchase3 (models/new-purchase (read-instant-date "2021-05-20T10:22:22") 240M "Rio de Janeiro"))
(def purchase4 (models/new-purchase (read-instant-date "2021-05-20T10:22:22") 540M "Rio de Janeiro"))
(def purchase5 (models/new-purchase (read-instant-date "2021-05-20T10:22:22") 540M "Rio de Janeiro"))
(def purchase6 (models/new-purchase (read-instant-date "2021-05-20T10:22:22") 540M "Rio de Janeiro"))

(db/create-purchase! conn purchase1 card1 category1)
(db/create-purchase! conn purchase2 card2 category2)
(db/create-purchase! conn purchase3 card3 category3)
(db/create-purchase! conn purchase4 card1 category2)
(db/create-purchase! conn purchase5 card3 category2)
(db/create-purchase! conn purchase6 card3 category2)
; -------------------- criação de compras -------------------

(pprint (db/todas-as-compras conn))
(pprint (db/top-purchases-count-client conn))
(pprint (db/top-purchases-value-client conn))
(pprint (db/client-without-purchase conn))

(db/apaga-banco!)
